### PR TITLE
Add blink::MemoryCache partitions for cross-site resource fetches.

### PR DESCRIPTION
**Pull request by goodov**
Resolves https://github.com/brave/brave-browser/issues/18831

| Before  | After |
| ------------- | ------------- |
| image | image |

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: chromium_src/third_party/blink/renderer/core/loader/frame_fetch_context.cc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include "../../../../../../../third_party/blink/renderer/core/loader/frame_fetch_context.cc"
+
+namespace blink {
+
+String FrameFetchContext::GetCacheIdentifierIfCrossSiteSubframe() const {
+  if (GetResourceFetcherProperties().IsDetached())
+    return cache_identifier_if_cross_site_subframe_;
+
+  String cache_identifier;
+  if (document_->domWindow()->IsCrossSiteSubframeIncludingScheme()) {
```
Comment (goodov) [inline chromium_src/third_party/blink/renderer/core/loader/frame_fetch_context.cc:15]: > `IsCrossSiteSubframeIncludingScheme()` uses HTML's definition of "site" as a registrable domain and scheme.
https://github.com/chromium/chromium/blob/98.0.4695.1/third_party/blink/renderer/core/frame/local_dom_window.h#L415-L416

I this this is very similar to what we do in `CookieSettingsBase::ShouldUseEphemeralStorage`, thus we can use this method as is. But if I'm wrong please tell me.
cc @pes10k

**Review by pes10k — APPROVED** [review]
lgtm!

File: chromium_src/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "third_party/blink/renderer/platform/loader/fetch/resource_fetcher.h"
+
+#include "third_party/blink/public/common/features.h"
+#include "third_party/blink/renderer/platform/loader/fetch/memory_cache.h"
+
+#define DefaultCacheIdentifier                           \
+  DefaultCacheIdentifier().IsEmpty()                     \
+      ? GetCacheIdentifierIfCrossSiteSubframeOrDefault() \
```
Comment (iefremov) [inline chromium_src/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc:13]: how about adding a comment to minimise WTF-effect

**Review by iefremov — APPROVED** [review]

File: chromium_src/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc (hunk 2 above)
Comment (goodov) [reply chromium_src/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc]: added more comments in various parts of the PR. the WTF-level should be decreased.

**Review by mariospr — APPROVED** [review]
I have to say that the EphemeralStorage-related code is becoming a bit hard to read due to so many changes scattered/fragmented throughout multiple chromium_src overrides, but other than that (plus a couple of small comments below) the PR LGTM.

Ideally, it would be great if we could somehow get this changes grouped in a more cohesive manner in some Brave-specific class(es) or similar, but not sure how feasible that really is.

File: browser/ephemeral_storage/ephemeral_storage_browsertest.cc [hunk 3]
```diff
@@ -79,12 +79,20 @@ std::unique_ptr<HttpResponse> HandleFileRequestWithNetworkCookies(
...
-                            base::SPLIT_WANT_ALL);
-      for (const auto& cookie : cookies) {
+      if (request_url.query() == "cache") {
         static_cast<BasicHttpResponse*>(http_response.get())
-            ->AddCustomHeader("Set-Cookie", cookie);
+            ->AddCustomHeader("Cache-Control",
+                              "public, max-age=604800, immutable");
+        static_cast<BasicHttpResponse*>(http_response.get())
+            ->AddCustomHeader("Etag", "etag");
+      } else {
```
[the remark this answers is not in the document]
Comment (goodov) [reply browser/ephemeral_storage/ephemeral_storage_browsertest.cc:87]: this whole block is inside `request_url.has_query()` check, so nothing has changed here.

**Comment by goodov** [comment]
> I have to say that the EphemeralStorage-related code is becoming a bit hard to read due to so many changes scattered/fragmented throughout multiple chromium_src overrides, but other than that (plus a couple of small comments below) the PR LGTM.
> 
> Ideally, it would be great if we could somehow get this changes grouped in a more cohesive manner in some Brave-specific class(es) or similar, but not sure how feasible that really is.

At some point we should be able to migrate to `blink::StorageKey` internals and this can help us simplify our changes. Currently it's a sad reality that a lot of parts are quite far logically and we don't have much tools to unite them in a nice way.

**Pull request merged**